src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +149,170 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+// to and including the OpaqueLoop* nodes). The cloned nodes are rewired to reflect the same graph structure as found for
+// this Template Assertion Predicate Expression. The cloned nodes get 'new_ctrl' as ctrl. There is no other update done
+// for the cloned nodes. Return the newly cloned Opaque4Node.
+Opaque4Node* TemplateAssertionPredicateExpression::clone(Node* new_ctrl, PhaseIdealLoop* phase) {
+  CloneStrategy clone_init_and_stride_strategy(phase, new_ctrl);
+  return clone(clone_init_and_stride_strategy, new_ctrl, phase);
+}
+
+// Class to collect data nodes from a source to target nodes by following the inputs of the source node recursively.
+// The class takes a node filter to decide which input nodes to follow and a target node predicate to start backtracking
+// from. All nodes found on all paths from source->target(s) returned in a Unique_Node_List (without duplicates).
+class DataNodesOnPathToTargets : public StackObj {
+  typedef bool (*NodeCheck)(const Node*);
+
+  NodeCheck _node_filter; // Node filter function to decide if we should process a node or not while searching for targets.
+  NodeCheck _is_target_node; // Function to decide if a node is a target node (i.e. where we should start backtracking).
+  Node_Stack _stack; // Stack that stores entries of the form: [Node* node, int next_unvisited_input_index_of_node].
+  VectorSet _visited; // Ensure that we are not visiting a node twice in the DFS walk which could happen with diamonds.
+  Unique_Node_List _collected_nodes; // The resulting node collection of all nodes on paths from source->target(s).
+
+ public:
+  DataNodesOnPathToTargets(NodeCheck node_filter, NodeCheck is_target_node)
+      : _node_filter(node_filter),
+        _is_target_node(is_target_node),
+        _stack(2) {}
+  NONCOPYABLE(DataNodesOnPathToTargets);
+
+  // Collect all input nodes from 'start_node'->target(s) by applying the node filter to discover new input nodes and
+  // the target node predicate to stop discovering more inputs and start backtracking. The implementation is done
+  // with an iterative DFS walk including a visited set to avoid redundant double visits when encountering a diamand
+  // shape which could consume a lot of unnecessary time.
+  const Unique_Node_List& collect(Node* start_node) {
+    assert(_collected_nodes.size() == 0, "should not call this method twice in a row");
+    assert(!_is_target_node(start_node), "no trivial paths where start node is also a target");
+
+    push_unvisited_node(start_node);
+    while (_stack.is_nonempty()) {
+      Node* current = _stack.node();
+      _visited.set(current->_idx);
+      if (_is_target_node(current)) {
+        // Target node? Do not visit its inputs and begin backtracking.
+        _collected_nodes.push(current);
+        pop_target_node_and_collect_predecessor();
+      } else if (!push_next_unvisited_input()) {
+        // All inputs visited. Continue backtracking.
```

Comment:
`push_next_unvisited_input`:
A `enum` with 2 values would make it more explicit than a bool: `HasMoreUnvisitedInputs` and `AllInputsAreVisited`.
Then it makes more sense at the use site, because you can just check for `AllInputsAreVisited`, and it is immediately clear that you can continue with backtacking.